### PR TITLE
[REEF-1310] The Java Driver should ACK the Java Evaluator's DONE hear…

### PR DESCRIPTION
**Pull request by afchung**
…tbeat

This addressed the issue by
  * Reflect the changes made in REEF-976 on the Java side.
  * For more details, see [REEF-976](https://issues.apache.org/jira/browse/REEF-976).

JIRA:
  [REEF-1310](https://issues.apache.org/jira/browse/REEF-1310)

### CODE REVIEW

**Comment by jwang98052** [comment]
Any way to add a test case for the scenario?

**Comment by afchung** [comment]
@jwang98052 I wasn't able to find an easy way to add tests that test the negative scenario without adding ugly hooks inside the Driver code to delay the `DONE` message. Would be great if you have ideas on how to do it.

**Comment by afchung** [comment]
This was originally discovered via a functional test for broadcast/reduce on the C# side, but the behavior was not deterministic either.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/evaluator/EvaluatorRuntime.java [hunk 1]
```diff
@@ -96,6 +96,14 @@ private void onEvaluatorControlMessage(final EvaluatorControlProto message) {
...
                 + "] sent to evaluator id[" + this.evaluatorIdentifier + "]"
         ));
+      } else if (ReefServiceProtos.State.DONE == this.state) {
+        if (message.getStopEvaluator() != null) {
+          LOG.log(Level.INFO, "Received ACK from Driver, shutting down Evaluator.");
+          this.clock.close();
+          return;
+        }
+
+        this.onException(new RuntimeException("Received a control message from Driver after Evaluator is done."));
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/evaluator/EvaluatorRuntime.java]: I personally would put this into the `else` statement.

**Comment by afchung** [comment]
@markusweimer I've addressed your comment. Also changed from `getStopEvaluator()` to `getDoneEvaluator()` for the `ACK` message.

**Comment by markusweimer** [comment]
LGTM, will merge.
